via-router/src/router.rs

```diff
@@ -43,17 +43,44 @@ impl<T> Node<T> {
 
 impl<T> Node<T> {
     #[inline]
-    pub fn param(&self) -> Option<&Param> {
-        match &self.pattern {
-            Pattern::Dynamic(param) | Pattern::Wildcard(param) => Some(param),
-            _ => None,
+    pub fn param<F>(&self, f: F) -> Option<(Param, [usize; 2])>
+    where
+        F: FnOnce() -> Option<[usize; 2]>,
+    {
+        if let Pattern::Dynamic(name) | Pattern::Wildcard(name) = &self.pattern {
+            Some((name.clone(), f()?))
+        } else {
+            None
         }
     }
 
     #[inline]
     pub fn route(&self) -> slice::Iter<MatchCond<T>> {
         self.route.iter()
     }
+
+    #[inline]
+    pub fn matches<'node, 'predicate, U>(
```

Comment:
This MatchCond enum is way to fun to play around with. In practice, it's probably better to unwrap the contained ref to node if you use params like we do in `AppService::call`.